internal/server/singleprocess/service_exec.go

```diff
@@ -51,10 +55,118 @@ func (s *service) StartExecStream(
 		}
 	case *pb.ExecStreamRequest_Start_DeploymentId:
 		log = log.With("deployment_id", t.DeploymentId)
-		err = s.state.InstanceExecCreateByDeployment(t.DeploymentId, execRec)
+
+		deployment, err := s.state.DeploymentGet(&pb.Ref_Operation{
+			Target: &pb.Ref_Operation_Id{
+				Id: t.DeploymentId,
+			},
+		})
 		if err != nil {
 			return err
 		}
+
+		// We need to spawn a job that will in turn spawn a virtual CEB
+		// that will connect back and create an instance exec record for us
+		// to use.
+		if deployment.HasExecPlugin {
+			instId, err := server.Id()
+			if err != nil {
+				return err
+			}
+
+			log.Info("spawning exec plugin via job", "instance-id", instId)
+
+			job := &pb.Job{
+				Workspace:   deployment.Workspace,
+				Application: deployment.Application,
+				Operation: &pb.Job_Exec{
+					Exec: &pb.Job_ExecOp{
+						InstanceId: instId,
+						Deployment: deployment,
+					},
+				},
+			}
+
+			// Means the client WANTS the job run on itself, so let's target the
+			// job back to it.
+			if runnerId, ok := server.RunnerId(srv.Context()); ok {
+				job.DataSource = &pb.Job_DataSource{
+					Source: &pb.Job_DataSource_Local{
+						Local: &pb.Job_Local{},
+					},
+				}
+
+				job.TargetRunner = &pb.Ref_Runner{
+					Target: &pb.Ref_Runner_Id{
+						Id: &pb.Ref_RunnerId{
+							Id: runnerId,
+						},
+					},
+				}
+
+				// Otherwise, the client wants an exec session but doesn't have a runner
+				// to use, so we just target any runner.
+			} else {
+				job.TargetRunner = &pb.Ref_Runner{
+					Target: &pb.Ref_Runner_Any{
+						Any: &pb.Ref_RunnerAny{},
+					},
+				}
+
+				// We leave DataSource unset here so that QueueJob will port over the data
+				// source from the project.
+			}
+
+			qresp, err := s.QueueJob(srv.Context(), &pb.QueueJobRequest{
+				Job: job,
+
+				// TODO unknown if this is enough time for when the request is queued
+				// by a runner-less client but a user waiting 60 seconds will get impatient
+				// regardless.
+				ExpiresIn: "60s",
+			})
+			if err != nil {
+				return err
+			}
```

Comment:
Do we need to defer and cancel the job? If this function exits, we want the job to stop right?

Reply:
We do that below on line 140, canceling it if waiting had an error. I wasn't sure about the semantic around canceling a completed job so I only did it where I knew I wanted it canceled.